### ISSUE TITLE
Isolation Forest for Anomaly Detection

### ISSUE THREAD

**Issue by iefymenkoanna** [ZDDduesseldorf/audioexplorer-backend#40]
## Code Review 

The first test version was: 
[audio-anomaly-detection.zip](https://github.com/user-attachments/files/28341551/audio-anomaly-detection.zip)

The actual version is on GitHub. 

## Summary
Implement Isolation Forest to calculate a global anomaly score for each audio embedding. The algorithm should be applied after the embedding vectors have been generated.

This issue documents the implementation of anomaly detection using Isolation Forest on audio embeddings.

The implementation operates directly on an embedding dictionary and does not require API integration.

The anomaly detection is executed through the `calculate_anomalies()` function and returns anomaly scores together with human-readable labels.

---

## Goal

Detect unusual audio samples based on their embeddings.
Calculate anomaly score for each audio sample.

The implementation should:

- receive embeddings as input
- execute Isolation Forest on all embeddings
- calculate anomaly scores
- convert scores into anomaly labels
- return structured results for each audio sample

---

## Description
Each audio sample is represented as a numerical embedding vector. Isolation Forest will analyze these vectors (Python) and assign an anomaly score indicating how unusual each sample is compared to the entire dataset.
The resulting score should be normalized and displayed in the Web UI as a percentage.

## Input Structure

The function receives a dictionary containing embeddings.

Example:

```python
{
    "uuid1": [...],
    "uuid2": [...],
    "uuid3": [...]
}
```

Where:

- key = audio UUID
- value = embedding vector

---

## Implementation

Main entry point:

```python
calculate_anomalies()
```

Workflow:

1. Receive embedding dictionary
2. Extract embedding vectors
3. Execute Isolation Forest
4. Calculate anomaly scores
5. Convert scores into labels
6. Return structured result dictionary

---

## Output Structure

The anomaly detection service returns a dictionary containing scores and labels for each audio sample.

Example:

```python
{
    "uuid1": {
        "scores": {
            "isolation_forest": 34.87,
            "lof": 78.77
        },
        "labels": {
            "isolation_forest": "Slightly Anomalous",
            "lof": "Anomalous"
        }
    }
}
```

--- 

## sklearn 

We use ensemble-learning algorithm. 

The implementation uses:

```python
from sklearn.ensemble import IsolationForest
```

Isolation Forest detects anomalies by isolating observations through recursive partitioning.

Samples requiring fewer partitions are considered more anomalous.

---

## Configuration

We need important configurations for our algorithm. 

### "contamination"
+ Controls how difficult it is for a data point to be classified as an anomaly.
+ Represents the expected proportion of anomalies in the dataset.
+ We could assume that approximately 3% of samples are anomalous.
      
### "n_estimators"
+ Number of decision trees used in the Isolation Forest ensemble.
+ More trees generally improve stability but increase computation time.
   
### "random_state"
+ Random seed used to ensure reproducible results across runs.
+ Using the same seed produces the same random behavior every time.

---

## Label Assignment

Raw scores are converted into human-readable labels
(based on their previous anomaly score). 

Labels: 

+ Not Anomalous (0-30%)
+ Slightly Anomalous (30-50%)
+ Anomalous (50-80%)
+ Highly Anomalous (80-100%)

Responsibilities:

- interpret anomaly scores
- assign anomaly categories
- provide consistent labels

---

## Unit Tests

Tests verify:

- processing of embedding dictionaries
- Isolation Forest execution
- anomaly score calculation
- label assignment
- output structure validation
- edge cases and boundary values

---

## Pipeline Position

Audio
→ Feature Extraction
→ CNN / Audio Encoder
→ Embeddings (numeric vectors)
→ Isolation Forest 
→ LOF 
→ Anomaly Score (%), Anomaly Label 
→ UMAP 
→ Web UI 

---

## UI Requirements
Display the score in the Web UI using the algorithm name:
Isolation Forest: 89.78% (example)
Label: Highly Anomalous (example)

Optional:
+ Tooltip: What is an Isolation Forest algorithm?
+ Tooltip text should explain that this is an unsupervised algorithm.

+ Diagrams (Bar chart). 

---

## Notes
+ Works directly on embedding vectors.
+ Does not require labels for the input. 
+ Detects global anomalies. 
+ Unsupervised learning approach.

--- 

## Unit Tests

Tests verify:

- processing of embedding dictionaries
- Isolation Forest execution
- anomaly score calculation
- label assignment
- output structure validation
- edge cases and boundary values

---

## Acceptance Criteria

- [x] Isolation Forest integrated
- [x] Embedding dictionary accepted as input
- [x] Scores calculated successfully
- [x] Labels assigned correctly
- [x] Output structure matches specification
- [x] Unit tests passing
- [x] Ruff checks passing
- [x] MyPy checks passing
- [x] CI pipeline passing
